hugo: helpers/content.go: call rst2html directly to render content

shreyanshk opened the pull request:
###### Motivation for this change

Hugo launches rst2html by calling Python interpreter and passing it the path of rst2html file.
This doesn't work on NixOS because the real rst2html is actually wrapped behind a bash script which is responsible for correctly setting the env and then exec-ing the real rst2html.

The problem is that rst2html is actually a bash script and not a Python script. Hence, Python fails to launch it with SyntaxError.

This patch ensures that rst2html is called directly so the shebang work correctly.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

Mic92 commented on pkgs/applications/misc/hugo/010-call-rst2html-directly.patch:
```diff
@@ -0,0 +1,51 @@
+ }
+ 
+-func getPythonExecPath() string {
+-	path, err := exec.LookPath("python")
```
This could be even incorrect if `python` is not the correct version python version for `rst2html`. Can you also open an issue upstream? They really only should do this on windows in my opinion and even there this probably also will work if rst2html is associated with the correct extension (i.e. rst2html.py).

Mic92 replied:
Also they check if `rst2html` is in `$PATH` and is executable. Is this even true on windows?

Mic92 replied:
After digging in LookPath on windows it appears that .py has to be in `PATHEXT` to be found by LookPath. This means if it is not already a registered extension it would not be found anyway. If it is registered then prepending python should be not necessary.

shreyanshk replied:
I know. I am unable to follow their reasoning as well.
I opened an issue upstream like a month ago but haven't received any response at all.
I think it's maybe because Nix is not on their priority list as they have quite a many other issues open. So, maybe we can apply the patch ourselves.

Of all the platforms where Nix works, except Windows, all natively support shebangs.
And, on Windows, Nix works under WSL (experimental) which supports shebangs. So, I believe, we can directly call rst2html for our use case without any issues.

shreyanshk replied:
Additionally for reference: https://github.com/gohugoio/hugo/issues/5167
Hopefully the devs will find a solution.

Mic92 replied:
Ok. Please leave the pull request url as commend above the patch and we go merging it until a better solution is found.

shreyanshk replied:
Just did that.
I think I also found a solution (https://github.com/gohugoio/hugo/pull/5285).

shreyanshk replied:
Ideally, it should be merged. Hugo's next release is at least a month away.

Mic92 replied:
The idea was to use `fetchpatch` instead.

shreyanshk replied:
Oh! let's wait for upstream to merge it.

Mic92 commented on pkgs/applications/misc/hugo/default.nix:
```diff
@@ -12,6 +12,9 @@ buildGoPackage rec {
     sha256 = "0n27vyg66jfx4lwswsmdlybly8c9gy5rk7yhy7wzs3rwzlqv1jzj";
   };
+  patches = [
+    ./010-call-rst2html-directly.patch
```
Ok. Now this can fetched from the upstream repository.

shreyanshk committed:
hugo: apply upstream patch to call rst2html directly on Nix

Initially, rst2html was called via the python interpreter which would
fail if the script was wrapped in a launcher as on NixOS.

Mic92 commented on pkgs/applications/misc/hugo/rst.nix:
```diff
@@ -0,0 +1,24 @@
+{ stdenv, hugo, python3, which }:
+
+stdenv.mkDerivation rec {
+  name = "hugoWithRST";
```
Do we really need this as a separate package? All it does is adding `pygments`. Users who need hugo + pygments can just install both or write a `shell.nix` for their project.

shreyanshk replied:
I tried. Installing both packages doesn't work. Hugo is able to find docutils but not pygments.
What I did was:
```
nix-env -iA nixpkgs.hugo
nix-env -iA nixpkgs.python36Packages.docutils
nix-env -iA nixpkgs.python36Packages.pygments
```
But, that resulted in hugo reporting that it couldn't find pygments.

It looks like docutils needs pygments and it is unable to find it even if both are installed (but separately) because pygments is not on docutils's PATH (or PYTHONPATH).

Additionally, sure shell.nix works but then shell customization are lost. So, I added it.

Mic92 replied:
I usually use direnv to preserve my shell: https://github.com/direnv/direnv/wiki/Nix

Mic92 replied:
It could be made a build option of hugo or a wrapper like weechat, but I would not clutter all-packages.nix with all possible dependency combinations.

shreyanshk replied:
This approach is certainly simpler. I'll remove the extra package. Thank you.

shreyanshk replied:
@Mic92 , I've updated the PR. Thanks.

shreyanshk commented:
@Mic92 ,
When would this patch be added into the nixos-18.09 channel? I see the channel was updated some time ago but the patch is still not present into that branch.
Thanks.

Mic92 commented:
I have not backported the commit back then to 18.09. This is now the case: f8847fc824664abcfa247a5b1ab78b877870f945
